src/m1_facilitator/handlers/GatewayProvenHandler.ts

```diff
@@ -0,0 +1,112 @@
+
+  public readonly remoteGatewayAddress: string;
+
+  public readonly blockNumber: BigNumber;
+
+  public constructor(
+    gatewayAddress: string,
+    remoteGatewayAddress: string,
+    blockNumber: number,
+  ) {
+    assert(Web3Utils.isAddress(gatewayAddress));
+    assert(Web3Utils.isAddress(remoteGatewayAddress));
+
+    this.gatewayAddress = Web3Utils.toChecksumAddress(gatewayAddress);
+    this.remoteGatewayAddress = Web3Utils.toChecksumAddress(remoteGatewayAddress);
+    this.blockNumber = new BigNumber(blockNumber);
+  }
+}
+
+/**
+ * GatewayProvenHandler class handles gateway-proven subgraph entities.
+ */
+export default class GatewayProvenHandler {
+  private readonly gatewayRepository: GatewayRepository;
+
+  /** Constructs GatewayProvenHandler from the specified arguments. */
+  public constructor(gatewayRepository: GatewayRepository) {
+    this.gatewayRepository = gatewayRepository;
+  }
+
+  /**
+   * handler() function accepts an array of records of a proven-gateway
+   * subgraph type.
+   *
+   * @pre There should exist a record in the `gatewayRepository` for each record
+   *      matching to `record.contractAddress` (in this case gateway address).
+   *
+   * @post Function updates a record of `gatewayRepository` matching to
+   *       `record.contractAddress` (gateway address) by setting
+   *       `remoteGatewayLastProvenBlockNumber` to the `record.blockNumber`
+   *        if `record.blockNumber` is greater than `remoteGatewayLastProvenBlockNumber`
+   *        of the stored value in the repository.
+   */
+  public async handle(records: any[]): Promise<void> {
+    const savePromises = records.map(async (record): Promise<void> => {
+      const gatewayProvenEntity = new GatewayProvenEntity(
+        record.contractAddress,
+        record.remoteGateway,
+        record.blockNumber,
+      );
+
+      const gatewayModelRecord: Gateway | null = await this.gatewayRepository.get(
+        gatewayProvenEntity.gatewayAddress,
+      );
+      assert(
+        gatewayModelRecord !== null,
+        'There is no gateway model in the gateway repository '
+        + `matching to ${gatewayProvenEntity.gatewayAddress}.`,
+      );
+
+      // We can safely cast as an assertion above is excluding the null case.
+      const gatewayModel: Gateway = gatewayModelRecord as Gateway;
+
+      const shouldUpdate = gatewayModel.remoteGatewayLastProvenBlockNumber === undefined
+        || gatewayModel.remoteGatewayLastProvenBlockNumber.isLessThan(
+          gatewayProvenEntity.blockNumber,
+        );
+
+      if (!shouldUpdate) {
+        return;
+      }
+
+      const updatedGatewayModel = new Gateway(
+        gatewayModel.gatewayGA,
+        gatewayModel.remoteGA,
+        gatewayModel.gatewayType,
+        gatewayModel.anchorGA,
+        gatewayProvenEntity.blockNumber,
+      );
+
+      await this.gatewayRepository.save(updatedGatewayModel);
+    });
+
+    await Promise.all(savePromises);
```

Comment:
what will happen when two Gateway proven entity records appear in handle method? 
I think, it will call save twice which intern will trigger services twice. It should only save once only the higher block height.

Reply:
That's a great point. The current implementation will fail. I've added mutex to guard that.

Reply:
I was proposing to only update block height(remoteGatewayLastProvenBlockNumber)  and ignore lower block height.

Reply:
Addressed (in a different way).